src/cockpit/bridge.py

```diff
@@ -159,6 +159,7 @@ def try_to_receive_stderr():
 
     try:
         stderr_fd, = fds
+        os.dup(2)
         os.dup2(stderr_fd, 2)
```

Comment:
I don't understand the "EVIL HACK" commit message: this dup2() would close the *bridge's* fd 2 (stderr) first, before setting it to the agent's stderr_fd -- at no point should it close stderr_fd here. I thought closing the latter would EOF the agent -- why would the agent care about the bridge's original stderr?

Reply:
The agent speaks to the subprocess via it's stderr and assumes that if stderr closes unexpectedly then the process has exited unexpectedly.

Reply:
OK, I apparently still misunderstand this. It would make sense to me if the cockpit-askpass subprocess sends it's fd 2 to the bridge, and the bridge's end is some random fd (stderr_fd). Connecting the bridge's fd 2 to the subprocess feels weird and wrong, as the bridge's stderr would/should either be unaltered (when calling from a terminal) or be connected to the journal (when running as part of the cockpit-session logind session).

(This is not to say "this is wrong", at this point it's "I don't understand what's going on")

Reply:
Ah, nevermind -- this is not about cockpit-bridge vs. askpass, but about the user's bridge vs. the --privileged bridge. I take it what happens is that unpriv-bridge lauches sudo which launches cockpit-bridge --privileged; and instead of just inheriting stderr (as this would  normally do), the unpriv bridge sends its stderr to the priv one, as it seems to assume that sudo would somehow mess up stderr? Does that ferny agent communication protocol somehow disrupt stderr? But then, what is the "agent"  that you were speaking about in this picture?

Reply:
Resolution: We have bridge (normal stderr) → sudo (redirected stderr) → bridge --priv (back to original normal stderr). So at this point in the code, stderr is still inherited from sudo/the agent, and this *restores* it to the original stderr fd that the unpriv bridge has.